tool-openssl/enc.cc

```diff
@@ -0,0 +1,202 @@
+    {"-e", kBooleanArgument, "Encrypt"},
+    {"-d", kBooleanArgument, "Decrypt"},
+    {"-K", kOptionalArgument, "Raw key to use, in hex form"},
+    {"-iv", kOptionalArgument, "IV to use, in hex form"},
+    {"-aes-128-cbc", kExclusiveBooleanArgument, "Supported cipher"},
+    {"", kOptionalArgument, ""}};
+
+static bool HexToBinary(uint8_t *buffer, const std::string &hex_string,
+                        unsigned int size) {
+  // First validate that the string contains only valid hex characters
+  for (char c : hex_string) {
+    if (!OPENSSL_isxdigit(c)) {
+      return false;
+    }
+  }
+
+  if (hex_string.size() != size * 2) {
+    return false;
+  }
+
+  BIGNUM *raw = NULL;
+  if (BN_hex2bn(&raw, hex_string.c_str()) == 0) {
+    return false;
+  }
+
+  int ret = BN_bn2binpad(raw, buffer, size);
+  BN_free(raw);
+  return ret != -1;
+}
+
+bool encTool(const args_list_t &args) {
+  ordered_args::ordered_args_map_t parsed_args;
+  args_list_t extra_args;
+  if (!ordered_args::ParseOrderedKeyValueArguments(parsed_args, extra_args,
+                                                   args, kArguments) ||
+      extra_args.size() > 0) {
+    PrintUsage(kArguments);
+    return false;
+  }
+
+  std::string in_path, out_path, hiv, cipher_name;
+  bssl::UniquePtr<std::string> hkey(new std::string);
+  bool help = false, e = false, d = false;
+
+  ordered_args::GetBoolArgument(&help, "-help", parsed_args);
+  ordered_args::GetString(&in_path, "-in", "", parsed_args);
+  ordered_args::GetString(&out_path, "-out", "", parsed_args);
+  ordered_args::GetBoolArgument(&e, "-e", parsed_args);
+  ordered_args::GetBoolArgument(&d, "-d", parsed_args);
+  ordered_args::GetString(hkey.get(), "-K", "", parsed_args);
+  ordered_args::GetString(&hiv, "-iv", "", parsed_args);
+  ordered_args::GetExclusiveBoolArgument(&cipher_name, kArguments, "",
+                                         parsed_args);
+
+  // Display enc tool option summary
+  if (help) {
+    PrintUsage(kArguments);
+    return true;
+  }
+
+  // Since we do not implement key generation, a raw key is required
+  // TODO: remove/modify if we ever implement -k, -kfile, or -S
+  if (hkey->empty()) {
+    fprintf(stderr, "Error: A raw key is required\n");
+    return false;
+  }
+
+  if (e && d) {
+    fprintf(stderr, "Error: -e and -d are mutually exclusive\n");
+    return false;
+  }
+
+  bool enc = true;
+  if (d) {
+    enc = false;
+  }
+
+  // Read from stdin if no -in path provided
+  ScopedFILE in_file;
+  if (in_path.empty()) {
+    in_file.reset(stdin);
+  } else {
+    in_file.reset(fopen(in_path.c_str(), "rb"));
+    if (!in_file) {
+      fprintf(stderr, "Error: unable to load data from '%s'\n",
+              in_path.c_str());
+      return false;
+    }
+  }
+
+  if (cipher_name.empty()) {
+    cipher_name = "aes-128-cbc";
+  } else {
+    cipher_name = cipher_name.substr(1);
+  }
+  const EVP_CIPHER *cipher = EVP_get_cipherbyname(cipher_name.c_str());
```

Comment:
Do we support more ciphers than `aes-128-cbc` here? That would be awesome, but it's a bit hard to tell from the help options above. Might want to make adjustments or document that we support more than just `aes-128-cbc` somewhere